addons/i18n_server/models/language_pack.py

```diff
@@ -0,0 +1,110 @@
+        "One pack can include more than one language (e.g. 'fr' includes 'fr.po' and fr_BE.po').",
+        required=True,
+    )
+    version = fields.Char(
+        help="Version of the Odoo server using the pack (e.g. '13.0', 'saas~14.3', '15.0alpha1')",
+        required=True,
+    )
+    description = fields.Text(help="Description of the archive (e.g. the module it contains)")
+    fname = fields.Char(string="Archive Filename", compute="_compute_display_name")
+
+    _sql_constraint = [
+        (
+            "unique_lang_version",
+            "UNIQUE(lang_code, version)",
+            "Only one language pack is possible per version",
+        )
+    ]
+
+    def _compute_display_name(self):
+        for pack in self:
+            pack.display_name = pack.version + " - " + pack.lang_code
+            pack.fname = pack.version + " - " + pack.lang_code + ".tar.xz"
+
+    def _generate_pack(self, languages, modules=None):
+        """ Generate i18n.packs
+
+        Based on the local po files inside the i18n/ directories
+        :param languages: a recordset of model res.lang, a pack will be generated using
+            the base language code (e.g. fr_FR and fr_BE will be in the same pack 'fr')
+        :param modules: a list of module name. If set, a pack will contain the
+            translation files for these modules only. If not set, all modules inside
+            the addons-path will be used.
+        """
+        langs = {lang.split("_")[0].split("@")[0] for lang in languages.mapped("code")}
+        version = release.version.split('+')[0]  # remove +e part if present
+
+        addons_path = {}
+        if not modules:
+            modules = get_modules()
+        for module in modules:
+            addons_path[module] = get_module_path(module)
+
+        for code in langs:
+            with tempfile.TemporaryFile(mode="w+b") as fileobj:
+                with tarfile.open(mode="x:xz", fileobj=fileobj) as tar:
+
+                    for module, path in addons_path.items():
+                        if not path:
+                            # invalid module
+                            continue
+
+                        i18n_path = os.path.join(path, "i18n", code + "*.po")
+                        for filename in glob.glob(i18n_path):
+                            # convert path to format '<module>/i18n/<lang>.po'
+                            # for file_open and use it as filename inside the archive
+                            local_path = "/".join(filename.split("/")[-3:])
+                            info = tarfile.TarInfo(local_path)
+                            with file_open(local_path, mode="rb") as file:
+                                # compute filesize
+                                file.seek(0, io.SEEK_END)
+                                info.size = file.tell()
+                                file.seek(0, io.SEEK_SET)
+
+                                tar.addfile(info, file)
```

Comment:
Does
```suggestion
                            local_path = "/".join(filename.split("/")[-3:])
                            tarfile.add(local_path)
```
not work?

Reply:
`AttributeError: module 'tarfile' has no attribute 'add'`

Reply:
Ah yes that should be `tar.add` not `tarfile.add`, got confused with the names.